File: C/Quick_sort.c
// بسم الله الرحمن الرحيم
// "وَأَنْ لَيْسَ لِلْإِنسَانِ إِلَّا مَا سَعَىٰ"
#include <stdio.h>
int Partition(int nums[], int l, int r)
{
    int pivot = nums[r];
    int i = l - 1, j = l;
    for (; j < r; j++)
    {
        if (nums[j] < pivot)
        {
            i++;
            int temp = nums[j];
            nums[j] = nums[i];
            nums[i] = temp;
        }
    }
    i++;
    int temp = nums[j];
    nums[j] = nums[i];
    nums[i] = temp;

    return i;
}

void Quick_sort(int nums[], int l, int r)
{

    if (l >= r)
        return;

    int pivot = Partition(nums, l, r);
    Quick_sort(nums, l, pivot - 1);
    Quick_sort(nums, pivot + 1, r);
}

int main()
{
    int n;
    scanf("%d", &n);
    int nums[n];
    for (int i = 0; i < n; i++)
        scanf("%d", &nums[i]);

    Quick_sort(nums, 0, n - 1);
    for (int i = 0; i < n; i++)
    {
        printf("%d ", nums[i]);
    }
    printf("\n");

    return 0;
}

File: Python/Quick_sort.py
# بسم الله الرحمن الرحيم
# "وَأَنْ لَيْسَ لِلْإِنسَانِ إِلَّا مَا سَعَىٰ"


# Time complexity: O(n log (N))
def Partition(nums, l, r):
    pivot = nums[r]
    i = l - 1
    j = l
    while j < r:
        if nums[j] < pivot:
            i += 1
            nums[i], nums[j] = nums[j], nums[i]
        j += 1
    i += 1
    nums[i], nums[j] = nums[j], nums[i]
    
    return i
         
    


def Quick_sort(nums, l, r):
    if l >= r:
        return
    pivot = Partition(nums, l, r)
    Quick_sort(nums, l, pivot - 1)
    Quick_sort(nums, pivot + 1, r)


def main():
    n = int(input())
    nums = list(map(int, input().split()))
    Quick_sort(nums, 0 , n - 1)
    print(*nums)


if __name__ == "__main__":
    main()
